extensions/pkg/util/cloudprofile.go

```diff
@@ -67,3 +73,117 @@ func NewV1beta1ImagesContext(parentImages []gardencorev1beta1.MachineImage) *Ima
 		},
 	)
 }
+
+// ValidateCapabilities validates the capabilities of a machine type or machine image against the capabilitiesDefinition
+// located in a cloud profile at spec.capabilities.
+// It checks if the capabilities are supported by the cloud profile and if the architecture is defined correctly.
+// It returns a list of field errors if any validation fails.
+func ValidateCapabilities(capabilities core.Capabilities, capabilitiesDefinitions []core.CapabilityDefinition, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// create map from capabilitiesDefinitions
+	capabilitiesDefinition := make(map[string][]string)
+	for _, capabilityDefinition := range capabilitiesDefinitions {
+		capabilitiesDefinition[capabilityDefinition.Name] = capabilityDefinition.Values
+	}
+	supportedCapabilityKeys := slices.Collect(maps.Keys(capabilitiesDefinition))
+
+	// Check if all capabilities are supported by the cloud profile
+	for capabilityKey, capability := range capabilities {
+		supportedValues, keyExists := capabilitiesDefinition[capabilityKey]
+		if !keyExists {
+			allErrs = append(allErrs, field.NotSupported(fldPath, capabilityKey, supportedCapabilityKeys))
+			continue
+		}
+		for i, value := range capability {
+			if !slices.Contains(supportedValues, value) {
+				allErrs = append(allErrs, field.NotSupported(fldPath.Child(capabilityKey).Index(i), value, supportedValues))
+			}
+		}
+	}
+
+	// Check additional requirements for architecture
+	// must be defined when multiple architectures are supported by the cloud profile
+	supportedArchitectures := capabilitiesDefinition[v1beta1constants.ArchitectureName]
+	architectures := capabilities[v1beta1constants.ArchitectureName]
+	if len(supportedArchitectures) > 1 && len(architectures) != 1 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child(v1beta1constants.ArchitectureName), architectures, "must define exactly one architecture when multiple architectures are supported by the cloud profile"))
+	}
+
+	return allErrs
+}
+
+// GetVersionCapabilitySets returns the CapabilitySets for a given machine image version or adds the default capabilitySet if non is defined.
+func GetVersionCapabilitySets(version core.MachineImageVersion, capabilitiesDefinitions []core.CapabilityDefinition) []core.CapabilitySet {
+	versionCapabilitySets := version.CapabilitySets
+	if len(version.CapabilitySets) == 0 {
+
```

Comment:
```suggestion
	versionCapabilitySets := version.CapabilitySets
	if len(versionCapabilitySets) == 0 {
```

Nit: If defined already, why not use it?

Reply:
To remove the indentation here, you could also short-circuit (with the added overhead of an additional return statement) and check whether `len(versionCapabilitySets) > 0` to return directly, else fill and return the fallback.